### PR TITLE
Serde support for revault transactions

### PR DESCRIPTION
**Pull request by JSwambo**
Based on https://github.com/re-vault/revault_tx/pull/21

Implements Serialize and Deserialize traits for revault transactions. Any data format (JSON, Bincode, CBOR) can be serialized into and out of by choosing the appropriate serde library.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -9,6 +9,8 @@ edition = "2018"
 bitcoin = "0.24"
 bitcoinconsensus = "0.19.0-1"
 miniscript = { git = "https://github.com/rust-bitcoin/rust-miniscript", branch = "master", features = ["compiler"] }
+serde = "1.0"
+serde_json = "1.0.58"
```
Comment (darosior) [inline Cargo.toml:13]: Can we make these feature-gated ?

**Comment by darosior** [comment]
Rebased on master.

File: src/transactions.rs [hunk 2]
```diff
@@ -1151,5 +1222,11 @@ mod tests {
         cancel_tx.hex().expect("Hex repr cancel_tx");
         emergency_tx.hex().expect("Hex repr emergency_tx");
         feebump_tx.hex().expect("Hex repr feebump_tx");
+
+        // Test serialize method for a transaction
+        let serialized_tx = serde_json::to_string(&vault_tx).unwrap();
+        let deserialized_tx = serde_json::from_str(&serialized_tx).unwrap();
+
+        assert_eq!(vault_tx, deserialized_tx);
```
Comment (darosior) [inline src/transactions.rs:1230]: I was thinking of a new test function doing roundtrips (ser / deser) for all transactions

File: src/transactions.rs [hunk 3]
```diff
@@ -369,6 +372,73 @@ macro_rules! impl_revault_transaction {
...
         }
+
+        impl Serialize for $transaction_name {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: Serializer,
+            {
+                let mut buff = Vec::<u8>::new();
+                match self.inner_tx().consensus_encode(&mut buff) {
+                    Ok(_) => {
```
Comment (darosior) [inline src/transactions.rs:383]: Maybe reuse `as_psbt_serialized()` ?

File: src/transactions.rs [hunk 4]
```diff
@@ -369,6 +372,73 @@ macro_rules! impl_revault_transaction {
...
+                        }
+
+                        Ok(vec)
+                    }
+                }
+
+                // Turn this into a rust vector with deserialize_seq()
+                let vec: Vec<u8> = deserializer
+                    .deserialize_seq(RevaultTxVisitor)
+                    .expect("Deserialization Error");
```
Comment (darosior) [inline src/transactions.rs:430]: Return the deserialization here instead ?

File: src/transactions.rs (hunk 2 above)
Comment (JSwambo) [reply src/transactions.rs:1230]: yes, will do that

**Comment by darosior** [comment]
Would be cool to add some cross-check against `bitcoind`'s `converttopsbt` output too (you can use `hex()` to generate test sample. Just as sanity checks as it turns out the roundtrips were not enough!

**Comment by darosior** [comment]
Just force-pushed: I simplified the existing test and added a sanity one against `bitcoind` psbt generation.

**Review by darosior — APPROVED** [review]
ACK 1aea165f229e919b47ae5a6e4fa715a7eea1dad8

**Pull request merged**